src/query/interpreter.cpp

```diff
@@ -2284,6 +2352,59 @@ PreparedQuery PrepareIsolationLevelQuery(ParsedQuery parsed_query, const bool in
       RWType::NONE};
 }
 
+Callback SwitchMemoryDevice(storage::StorageMode current_mode, storage::StorageMode requested_mode,
+                            InterpreterContext *interpreter_context) {
+  Callback callback;
+  callback.fn = [current_mode, requested_mode, interpreter_context]() mutable {
+    if (current_mode == requested_mode) {
+      return std::vector<std::vector<TypedValue>>();
+    }
+    if (SwitchingFromDiskToInMemory(current_mode, requested_mode)) {
+      throw utils::BasicException(
+          "You cannot switch from on-disk storage to in-memory storage while the database is running. "
+          "Please delete your data directory and restart the database. Once restarted, the Memgraph will automatically "
+          "start to use in-memory storage.");
+    }
+    if (SwitchingFromInMemoryToDisk(current_mode, requested_mode)) {
+      std::unique_lock main_guard{interpreter_context->db->main_lock_};
+
+      if (auto vertex_cnt_approx = interpreter_context->db->GetInfo().vertex_count; vertex_cnt_approx > 0) {
+        throw utils::BasicException(
+            "You cannot switch from in-memory storage to on-disk storage when the database "
+            "contains data. Please delete all entries from the database, run FREE MEMORY and then repeat this "
+            "query. ");
+      }
+
+      main_guard.unlock();
```

Comment:
lock must be released before new DiskStorage has been created. And yes you are right that some thread can access the DB while being switched. However, this is solved by having creation_storage_mode_ in the accessor.  Then upon commit, I just check whether the storage mode is the same as during the start

Reply:
But what happens to the storage if someone else is using it while you are switching it?
The `interpreter_context->db = std::make_unique<memgraph::storage::DiskStorage>(db_config);` will destroy the old storage. We have to make sure no one else can possibly be using it while this is happening. 

Nothing is stopping someone from accessing the original storage between lines:
`if (interpreter_context->interpreters->size() > 1) {`
and
`interpreter_context->db = std::make_unique<memgraph::storage::DiskStorage>(db_config);`.
I imagine that the DiskStorage creation can take some time, so it's not impossible that this happens.
If the only check is on commit, then someone could be using the destroyed storage for a long time.

Reply:
we don't know what happens, probably it would but the pure point of switching to disk is that you should it alone, at the beginning. So yes there is probably a moment in which something would crash but I wouldn't say the problem is someone is using destroyed storage before commit. But again I agree this should be somehow refactored but this is not easy part. Maybe when we tackle the issue of switching from Lab

Reply:
Maybe something like this could work.
Every new connection adds an interpreter to the list. So this way they can't use it until we moved the storage.
**NOTE:** No idea if this code is actually okay.
```
interpreter_context->interpreters.WithLock([&](const auto &interpreters_) {
  if (interpreters_.size() > 1) {
    throw utils::BasicException(
        "You cannot switch from an in-memory storage mode to the on-disk storage mode when there are "
        "multiple sessions active. Close all other sessions and try again. As Memgraph Lab uses "
        "multiple sessions to run queries in parallel,  "
        "it is currently impossible to switch to the on-disk storage mode within Lab. "
        "Close it, connect to the instance with mgconsole "
        "and change the storage mode to on-disk from there. Then, you can reconnect with the Lab "
        "and continue to use the instance as usual.");
  }
  main_guard.unlock();
  interpreter_context->db = std::make_unique<memgraph::storage::DiskStorage>(interpreter_context->db->config_);
});
```